File: terra/src/main/scala/com/dafttech/terra/game/world/entities/EntityDiggerBeam.java
package com.dafttech.terra.game.world.entities;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.engine.lighting.PointLight;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.entities.models.EntityThrown;
import com.dafttech.terra.game.world.tiles.Tile;
import com.dafttech.terra.resources.Options;
import com.dafttech.terra.resources.Resources$;

import java.util.Random;

public class EntityDiggerBeam extends EntityThrown {
    PointLight light;

    public EntityDiggerBeam(Vector2 pos, World world) {
        super(pos, world, new Vector2(4f, 2f));
        setGravityFactor(0);

        isDynamicEntity = true;
    }

    @Override
    public TextureRegion getImage() {
        return Resources$.MODULE$.ENTITIES().getImage("beamDig");
    }

    @Override
    public void update(World world, float delta) {
        super.update(world, delta);

        if (light == null) {
            light = new PointLight(getPosition(), 95);
            light.setColor(new Color(0, 1, 0.3f, 1));
        }

        light.setSize(90 + new Random().nextInt(10));

        light.setPosition(getPosition().$plus(size.x() * Options.BLOCK_SIZE() / 2, size.y() * Options.BLOCK_SIZE() / 2));

        if (Math.abs(velocity.x()) <= 0.1 && Math.abs(velocity.y()) <= 0.1) {
            worldObj.removeEntity(this);
        }
    }

    @Override
    public boolean collidesWith(Entity e) {
        return false;
    }

    @Override
    public void onTerrainCollision(Tile tile) {
        if (!tile.isAir()) {
            worldObj.destroyTile(tile.getPosition().x(), tile.getPosition().y(), this).addVelocity(velocity.$times(-1));
            this.remove();
        }
    }

    @Override
    public float getInAirFriction() {
        return 0;
    }

    @Override
    public boolean isLightEmitter() {
        return true;
    }

    @Override
    public PointLight getEmittedLight() {
        return light;
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/entities/EntityDiggerBeam.scala
package com.dafttech.terra.game.world.entities

import java.util.Random

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.engine.lighting.PointLight
import com.dafttech.terra.engine.vector.{Vector2d, Vector2i}
import com.dafttech.terra.game.world.GameWorld
import com.dafttech.terra.game.world.entities.models.EntityThrown
import com.dafttech.terra.resources.Resources
import monix.eval.Task

class EntityDiggerBeam(pos: Vector2d)(implicit world: GameWorld) extends EntityThrown(pos) {
  private val light: PointLight = new PointLight(6)
  light.setColor(new Color(0, 1, 0.3f, 1))

  override def getImage: Task[TextureRegion] = Resources.ENTITIES.getImageTask("beamDig")

  override def update(delta: Float)(implicit tilePosition: TilePosition): Unit = {
    super.update(delta)

    light.setSize(90 + new Random().nextInt(10))

    if (Math.abs(body.linVelWorld.x) <= 0.1 && Math.abs(body.linVelWorld.y) <= 0.1)
      world.removeEntity(this)
  }

  override def onTerrainCollision(t: TilePosition): Unit = if (!t.getTile.isAir) {
    world.destroyTile(Vector2i(t.pos.x, t.pos.y), this)
      //.foreach(_.addVelocity(velocity.$times(-1)))
    world.removeEntity(this)
  }

  override def getEmittedLight: PointLight = light
}